src/XMakeTasks/FileState.cs

```diff
@@ -36,33 +35,213 @@ namespace Microsoft.Build.Tasks
+                        LastWriteTimeUtc =
+                            DateTime.FromFileTimeUtc(((long)data.ftLastWriteTimeHigh << 0x20) | data.ftLastWriteTimeLow);
+                        Length = IsDirectory ? 0 : (((long)data.fileSizeHigh << 0x20) | data.fileSizeLow);
+                    }
+                    else
+                    {
+                        // Check if we have a directory
+                        IsDirectory = Directory.Exists(_filename);
+                        Exists = IsDirectory;
+
+                        // If not exists, see if this is a file
+                        if (!Exists)
+                        {
+                            Exists = File.Exists(_filename);
+                        }
+
+                        if (IsDirectory)
+                        {
+                            // Use DirectoryInfo to get the last write date
+                            var directoryInfo = new DirectoryInfo(_filename);
+                            IsReadOnly = false;
+                            LastWriteTimeUtc = directoryInfo.LastWriteTimeUtc;
+                        }
+                        else if (Exists)
+                        {
+                            // Use FileInfo to get readonly and last write date
+                            var fileInfo = new FileInfo(_filename);
+                            IsReadOnly = fileInfo.IsReadOnly;
+                            LastWriteTimeUtc = fileInfo.LastWriteTimeUtc;
+                        }
+                    }
+                }
+                catch (Exception ex)
+                {
+                    // Save the exception thrown and assume the file does not exist
+                    _exceptionThrown = ex;
+                    Exists = false;
+                }
+                finally
+                {
+                    // Reset the error mode on Windows
+                    if (NativeMethodsShared.IsWindows)
+                    {
+                        NativeMethodsShared.SetErrorMode(oldMode);
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Throw exception as if the FileInfo did it. We
+            /// know that getting the length of a file would
+            /// throw exception if there are IO problems
+            /// </summary>
+            /// <param name="doThrow"></param>
+            public void ThrowFileInfoException(bool doThrow)
+            {
+                if (doThrow)
+                {
+                    // Provoke exception
+                    var length = (new FileInfo(_filename)).Length;
+                }
+            }
+
+            /// <summary>
+            /// Throw non-IO-related exception. Return true if
+            /// exception exists, but was IO-related
+            /// </summary>
+            /// <returns></returns>
+            public bool ThrowNonIoException()
```

Comment:
Can you give this a name that makes it clearer that it does nothing if there's no exception to be thrown?  I got confused by some of the calling code.